staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go

```diff
@@ -0,0 +1,366 @@
+	endpointsSynced cache.InformerSynced
+
+	// To allow injection for testing.
+	syncFn func(key string) error
+
+	queue workqueue.RateLimitingInterface
+}
+
+func NewAvailableConditionController(
+	apiServiceInformer informers.APIServiceInformer,
+	serviceInformer v1informers.ServiceInformer,
+	endpointsInformer v1informers.EndpointsInformer,
+	apiServiceClient apiregistrationclient.APIServicesGetter,
+) *AvailableConditionController {
+	c := &AvailableConditionController{
+		apiServiceClient: apiServiceClient,
+		apiServiceLister: apiServiceInformer.Lister(),
+		apiServiceSynced: apiServiceInformer.Informer().HasSynced,
+		serviceLister:    serviceInformer.Lister(),
+		servicesSynced:   serviceInformer.Informer().HasSynced,
+		endpointsLister:  endpointsInformer.Lister(),
+		endpointsSynced:  endpointsInformer.Informer().HasSynced,
+		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "AvailableConditionController"),
+	}
+
+	apiServiceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addAPIService,
+		UpdateFunc: c.updateAPIService,
+		DeleteFunc: c.deleteAPIService,
+	})
+
+	serviceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addService,
+		UpdateFunc: c.updateService,
+		DeleteFunc: c.deleteService,
+	})
+
+	endpointsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addEndpoints,
+		UpdateFunc: c.updateEndpoints,
+		DeleteFunc: c.deleteEndpoints,
+	})
+
+	c.syncFn = c.sync
+
+	return c
+}
+
+func (c *AvailableConditionController) sync(key string) error {
+	inAPIService, err := c.apiServiceLister.Get(key)
+	if apierrors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	apiService := &apiregistration.APIService{}
+	if err := apiregistration.DeepCopy_apiregistration_APIService(inAPIService, apiService, cloner); err != nil {
+		return err
+	}
+
+	availableCondition := apiregistration.APIServiceCondition{
+		Type:               apiregistration.Available,
+		Status:             apiregistration.ConditionTrue,
+		LastTransitionTime: metav1.Now(),
+	}
+
+	// local API services are always considered available
+	if apiService.Spec.Service == nil {
+		availableCondition.Status = apiregistration.ConditionTrue
+		availableCondition.Reason = "Local"
+		availableCondition.Message = "Local APIServices are always available"
+		apiregistration.SetAPIServiceCondition(apiService, availableCondition)
+		_, err := c.apiServiceClient.APIServices().UpdateStatus(apiService)
+		return err
+	}
+
+	service, err := c.serviceLister.Services(apiService.Spec.Service.Namespace).Get(apiService.Spec.Service.Name)
+	if apierrors.IsNotFound(err) {
+		availableCondition.Status = apiregistration.ConditionFalse
+		availableCondition.Reason = "NoService"
+		availableCondition.Message = fmt.Sprintf("service/%s in %q is not present", apiService.Spec.Service.Name, apiService.Spec.Service.Namespace)
```

Comment:
true. Never liked the CLI pattern ;-) If it matches what we have there, it's fine.